phoenix-core/src/main/java/org/apache/phoenix/coprocessor/GlobalIndexRegionScanner.java

```diff
@@ -1330,24 +1345,43 @@ public static List<Mutation> prepareIndexMutationsForRebuild(IndexMaintainer ind
                 applyDeleteOnPut(deleteToApply, currentDataRowState);
                 Put nextDataRowState = currentDataRowState;
                 if (nextDataRowState.getFamilyCellMap().size() == 0) {
-                    Mutation del = indexMaintainer.buildRowDeleteMutation(indexRowKeyForCurrentDataRow,
-                            IndexMaintainer.DeleteType.ALL_VERSIONS, ts);
-                    indexMutations.add(del);
+                    if (indexRowKeyForCurrentDataRow != null) {
+                        Mutation
+                                del =
+                                indexMaintainer.buildRowDeleteMutation(indexRowKeyForCurrentDataRow,
+                                        IndexMaintainer.DeleteType.ALL_VERSIONS, ts);
+                        indexMutations.add(del);
+                    }
                     currentDataRowState = null;
                     indexRowKeyForCurrentDataRow = null;
-                } else {
+                } else if (indexRowKeyForCurrentDataRow != null) {
+                    if (!IndexRegionObserver.shouldPrepareIndexMutations(indexMaintainer,
+                            nextDataRowState)) {
+                        currentDataRowState = nextDataRowState;
+                        Mutation del = indexMaintainer.buildRowDeleteMutation(
+                                indexRowKeyForCurrentDataRow,
+                                IndexMaintainer.DeleteType.ALL_VERSIONS, ts);
+                        indexMutations.add(del);
+                        indexRowKeyForCurrentDataRow = null;
+                        continue;
+                    }
                     ValueGetter nextDataRowVG = new IndexUtil.SimpleValueGetter(nextDataRowState);
-                    Put indexPut = prepareIndexPutForRebuid(indexMaintainer, rowKeyPtr, nextDataRowVG, ts);
+                    Put indexPut = prepareIndexPutForRebuid(indexMaintainer, rowKeyPtr,
```

Comment:
is this a typo in function name **Rebuild**?

Reply:
Yes

Reply:
Done